ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h

```diff
@@ -0,0 +1,550 @@
+            local_Jac_data, local_matrix_size, local_matrix_size);
+
+        auto local_rhs = MathLib::createZeroedVector<RhsVector>(
+            local_rhs_data, local_matrix_size);
+
+        typename ShapeMatricesType::template MatrixType<displacement_size,
+                                                        temperature_size> KuT;
+        KuT.setZero(displacement_size, temperature_size);
+
+        typename ShapeMatricesType::NodalMatrixType KTT;
+        KTT.setZero(temperature_size, temperature_size);
+
+        typename ShapeMatricesType::NodalMatrixType DTT;
+        DTT.setZero(temperature_size, temperature_size);
+
+        double const& dt = _process_data.dt;
+
+        SpatialPosition x_position;
+        x_position.setElementID(_element.getID());
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            x_position.setIntegrationPoint(ip);
+            auto const& w = _ip_data[ip].integration_weight;
+
+            auto const& dNdx = _ip_data[ip].dNdx;
+            auto const& N = _ip_data[ip].N;
+
+            auto const& B = _ip_data[ip].b_matrices;
+            auto const& sigma = _ip_data[ip].sigma;
+
+            auto const& C = _ip_data[ip].C;
+
+            double const delta_T = N.dot(T) - _process_data.reference_temperature;
+            // calculate thermally induced strain
+            // assume isotropic thermal expansion
+            auto const alpha =
+                _process_data.linear_thermal_expansion_coefficient(
+                    t, x_position)[0];
+            double const linear_thermal_strain = alpha * delta_T;
+
+            //
+            // displacement equation, displacement part
+            //
+            _ip_data[ip].updateConstitutiveRelation(t, x_position, dt, u,
+                                                    linear_thermal_strain);
+
+            local_Jac.template block<displacement_size, displacement_size>(
+                         displacement_index, displacement_index)
+                .noalias() += B.transpose() * C * B * w;
+
+            typename ShapeMatricesType::template MatrixType<
+                DisplacementDim, displacement_size> N_u = ShapeMatricesType::
+                template MatrixType<DisplacementDim, displacement_size>::Zero(
+                    DisplacementDim, displacement_size);
+
+            for (int i = 0; i < DisplacementDim; ++i)
+                N_u.template block<1, displacement_size / DisplacementDim>(
+                       i, i * displacement_size / DisplacementDim)
+                    .noalias() = N;
+
+            using Invariants =
+                MaterialLib::SolidModels::Invariants<kelvin_vector_size>;
+
+            // calculate real density
+            auto const rho_sr = _process_data.solid_density(t, x_position)[0];
+            double const rho_s = rho_sr * (1 - 3 * linear_thermal_strain);
+
```

Comment:
@endJunction Here you see what's happening. rho_s is the current density, that's used in the assembly. It is calculated from the reference value at the reference temperature.